File: Web/src/service/AttendanceService.ts
import ApiService from './ApiService';

export interface AttendanceRecord {
  id: number;
  student_id: number;
  subject_id: number | null;
  teacher_id: number;
  schedule_id: number | null;
  date: string;
  time: string | null;
  status: 'present' | 'absent' | 'late' | 'excused';
  reason: string | null;
  created_at: string;
  updated_at: string;
  student?: {
    id: number;
    first_name: string;
    last_name: string;
    code: string;
  };
  subject?: {
    id: number;
    name: string;
  };
}

export interface CreateAttendanceDTO {
  student_id: number;
  subject_id?: number | null;
  teacher_id: number;
  schedule_id?: number | null;
  status: 'present' | 'absent' | 'late' | 'excused';
  reason?: string | null;
  date?: string | null;
}

export interface UpdateAttendanceDTO {
  status: 'present' | 'absent' | 'late' | 'excused';
  reason?: string | null;
}

class AttendanceService {
  async getClassAttendances(
    classId: number,
    params?: { date?: string; start_date?: string; end_date?: string; subject_id?: number; schedule_id?: number }
  ): Promise<AttendanceRecord[]> {
    const query: Record<string, any> = {};
    if (params?.date) {
      query.start_date = params.date;
      query.end_date = params.date;
    } else {
      if (params?.start_date) query.start_date = params.start_date;
      if (params?.end_date) query.end_date = params.end_date;
    }
    if (params?.subject_id) query.subject_id = params.subject_id;
    if (params?.schedule_id) query.schedule_id = params.schedule_id;

    const response = await ApiService.get<AttendanceRecord[]>(`/classes/${classId}/attendances`, query);
    return (response.data as any)?.data ?? response.data ?? [];
  }

  async getScheduleAttendances(
    scheduleId: number,
    params?: { date?: string; class_id?: number }
  ): Promise<AttendanceRecord[]> {
    const query: Record<string, any> = {};
    if (params?.date) query.date = params.date;
    if (params?.class_id) query.class_id = params.class_id;

    const response = await ApiService.get<AttendanceRecord[]>(`/schedules/${scheduleId}/attendances`, query);
    return (response.data as any)?.data ?? response.data ?? [];
  }

  async createAttendance(data: CreateAttendanceDTO): Promise<AttendanceRecord> {
    const response = await ApiService.post<AttendanceRecord>('/attendances', data);
    return (response.data as any);
  }

  async updateAttendance(id: number, data: UpdateAttendanceDTO): Promise<AttendanceRecord> {
    const response = await ApiService.put<AttendanceRecord>(`/attendances/${id}`, data);
    return (response.data as any);
  }

  /**
   * Save attendance for an entire class in a single bulk request.
   * existingMap: Map<student_id, AttendanceRecord> for that date/subject/schedule.
   */
  async saveClassAttendance(
    entries: Array<{
      student_id: number;
      subject_id: number | null;
      teacher_id: number;
      schedule_id?: number | null;
      status: 'present' | 'absent' | 'late' | 'excused';
      reason?: string | null;
      date?: string | null;
    }>,
    existingMap: Map<number, AttendanceRecord>
  ): Promise<void> {
    const records = entries.map((entry) => {
      const existing = existingMap.get(entry.student_id);
      if (existing) {
        return { id: existing.id, status: entry.status, reason: entry.reason ?? null };
      }
      return {
        student_id:  entry.student_id,
        subject_id:  entry.subject_id,
        teacher_id:  entry.teacher_id,
        schedule_id: entry.schedule_id ?? null,
        date:        entry.date ?? null,
        status:      entry.status,
        reason:      entry.reason ?? null,
      };
    });

    await ApiService.post('/attendances/bulk', { records });
  }
}

export default new AttendanceService();

File: Mobile/lib/services/attendance_service.dart
import '../models/attendance.dart';
import 'api_service.dart';

class AttendanceService {
  final ApiService _api;

  AttendanceService(this._api);

  /// Get attendance records for a student
  Future<List<AttendanceRecord>> getStudentAttendances(int studentId, {
    String? startDate,
    String? endDate,
    String? status,
    int? subjectId,
    int? month,
    String? semester,
  }) async {
    final params = <String, dynamic>{};
    if (startDate != null) params['start_date'] = startDate;
    if (endDate != null) params['end_date'] = endDate;
    if (status != null) params['status'] = status;
    if (subjectId != null) params['subject_id'] = subjectId;
    if (month != null) params['month'] = month;
    if (semester != null) params['semester'] = semester;

    final response = await _api.get(
      '/parent/students/$studentId/attendances',
      params: params.isNotEmpty ? params : null,
    );

    final data = _api.extractData(response);

    if (data is List) {
      return data
          .map((a) => AttendanceRecord.fromJson(a as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  /// Get attendance records for a class (teacher view)
  Future<List<AttendanceRecord>> getClassAttendances(int classId, {
    String? date,
    String? startDate,
    String? endDate,
    int? subjectId,
    int? scheduleId,
  }) async {
    final params = <String, dynamic>{};

    if (date != null) {
      params['start_date'] = date;
      params['end_date'] = date;
    } else {
      if (startDate != null) params['start_date'] = startDate;
      if (endDate != null) params['end_date'] = endDate;
    }
    if (subjectId != null) params['subject_id'] = subjectId;
    if (scheduleId != null) params['schedule_id'] = scheduleId;

    final response = await _api.get(
      '/classes/$classId/attendances',
      params: params.isNotEmpty ? params : null,
    );

    final data = _api.extractData(response);
    if (data is List) {
      return data
          .map((a) => AttendanceRecord.fromJson(a as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  /// Save attendance for a class in bulk
  Future<void> saveClassAttendance(
    List<Map<String, dynamic>> records,
  ) async {
    await _api.post('/attendances/bulk', data: { 'records': records });
  }
}
